common/functions/src/scalars/conditionals/if.rs

```diff
@@ -11,59 +11,144 @@
+            nullable = true;
+            let boolean_dt: DataTypePtr = Arc::new(NullableType::create(BooleanType::arc()));
+            predicate = cast_column_field(&columns[0], &boolean_dt)?;
+        } else {
+            let boolean_dt = BooleanType::arc();
+            predicate = cast_column_field(&columns[0], &boolean_dt)?;
+        }
+
+        let lhs = &columns[1];
+        let rhs = &columns[2];
+        let l_dt = remove_nullable(lhs.data_type());
+        let r_dt = remove_nullable(rhs.data_type());
+
+        let mut least_supertype = compare_coercion(&l_dt, &r_dt)?;
+
+        if (lhs.data_type().is_nullable() || rhs.data_type().is_nullable())
+            && !least_supertype.is_nullable()
+        {
+            least_supertype = Arc::new(NullableType::create(least_supertype));
+        }
+        if least_supertype.is_nullable() {
+            nullable = true;
+        }
+        let lhs = cast_column_field(lhs, &least_supertype)?;
+        let rhs = cast_column_field(rhs, &least_supertype)?;
+
+        let type_id = remove_nullable(&lhs.data_type()).data_type_id();
+
+        macro_rules! primitive_build {
+            (
+             $T:ident
+        ) => {{
+                let predicate_wrapper = ColumnViewer::<bool>::create(&predicate)?;
+                let lhs_wrapper = ColumnViewer::<$T>::create(&lhs)?;
+                let rhs_wrapper = ColumnViewer::<$T>::create(&rhs)?;
+                let size = lhs_wrapper.len();
+
+                if nullable {
+                    let mut builder = NullableColumnBuilder::<$T>::with_capacity(size);
+
+                    for row in 0..size {
+                        let predicate = predicate_wrapper.value(row);
+                        let valid = predicate_wrapper.valid_at(row);
+                        if predicate {
+                            builder
+                                .append(lhs_wrapper.value(row), valid & lhs_wrapper.valid_at(row));
+                        } else {
+                            builder
+                                .append(rhs_wrapper.value(row), valid & rhs_wrapper.valid_at(row));
+                        };
+                    }
+
+                    Ok(builder.build(size))
+                } else {
+                    let mut builder = ColumnBuilder::<$T>::with_capacity(size);
+
+                    for row in 0..size {
+                        let predicate = predicate_wrapper.value(row);
+                        if predicate {
+                            builder.append(lhs_wrapper.value(row));
+                        } else {
+                            builder.append(rhs_wrapper.value(row));
+                        };
+                    }
+
+                    Ok(builder.build(size))
+                }
+            }};
+        }
+
+        with_match_primitive_type2!(type_id, |$T| {
+            primitive_build!($T)
+        }, {
+            match type_id {
+                TypeID::String => primitive_build!(Vu8),
```

Comment:
with_match_primitive_type2 

=> with_match_scalar_types2  to cover these two?

Reply:
yes, resolved